File: 2226-maximum-candies-allocated-to-k-children/2226-maximum-candies-allocated-to-k-children.py
class Solution:
    def maximumCandies(self, candies: List[int], k: int) -> int:
        low, high = 0, max(candies)

        while low < high:
            mid = low + (high - low + 1) // 2
            if self.canGive(candies, k, mid):
                low = mid
            else:
                high = mid - 1
        return low

    def canGive(self, candies: List[int], k: int, count: int) -> bool:
        if count == 0: return True

        for candy in candies:
            k -= candy // count
            if k <= 0: return True

        return False

File: 2226-maximum-candies-allocated-to-k-children/2226-maximum-candies-allocated-to-k-children.cs
public class Solution {
    public int MaximumCandies(int[] candies, long k) {
        int low = 0, high = candies.Max();

        while(low < high){
            var mid = low + (high - low + 1) / 2;
            if(CanGive(candies, k, mid))
                low = mid;
            else 
                high = mid - 1;
        }

        return low;
    }

    private bool CanGive(int[] candies, long k, int count){
        if(count == 0) return true;

        for(var i=0; i<candies.Length && k > 0; i++)
            k -= candies[i] / count;

        return k <= 0;
    }
}